cmd/gitops/add/profiles/cmd.go

```diff
@@ -0,0 +1,132 @@
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+)
+
+var opts profiles.AddOptions
+
+// AddCommand provides support for adding a profile to a cluster.
+func AddCommand(client *resty.Client) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:           "profile",
+		Aliases:       []string{"profiles"},
+		Short:         "Add a profile to a cluster",
+		SilenceUsage:  true,
+		SilenceErrors: true,
+		Example: `
+		# Add a profile to a cluster
+		gitops add profile --name=podinfo --cluster=prod --version=1.0.0 --config-repo=ssh://git@github.com/owner/config-repo.git
+		`,
+		RunE: addProfileCmdRunE(client),
+	}
+
+	cmd.Flags().StringVar(&opts.Name, "name", "", "Name of the profile")
+	cmd.Flags().StringVar(&opts.Version, "version", "latest", "Version of the profile")
+	cmd.Flags().StringVar(&opts.ConfigRepo, "config-repo", "", "URL of external repository (if any) which will hold automation manifests")
+	cmd.Flags().StringVar(&opts.Cluster, "cluster", "", "Name of the cluster to add the profile to")
+	cmd.Flags().StringVar(&opts.Port, "port", server.DefaultPort, "Port the profiles API is running on")
+	cmd.Flags().BoolVar(&opts.AutoMerge, "auto-merge", false, "If set, 'gitops add profile' will merge automatically into the set --branch")
+
+	return cmd
+}
+
+func addProfileCmdRunE(client *resty.Client) func(*cobra.Command, []string) error {
+	return func(cmd *cobra.Command, args []string) error {
+		log := internal.NewCLILogger(os.Stdout)
+		fluxClient := flux.New(osys.New(), &runner.CLIRunner{})
+		factory := services.NewFactory(fluxClient, log)
+		providerClient := internal.NewGitProviderClient(os.Stdout, os.LookupEnv, auth.NewAuthCLIHandler, log)
+
+		if err := validateAddOptions(opts); err != nil {
+			return err
+		}
+
+		ns, err := cmd.Parent().Parent().Flags().GetString("namespace")
+		if err != nil {
+			return err
+		}
+		opts.Namespace = ns
+
+		config, err := clientcmd.BuildConfigFromFlags("", filepath.Join(homedir.HomeDir(), ".kube", "config"))
+		if err != nil {
+			return fmt.Errorf("error initializing kubernetes config: %w", err)
+		}
+
+		clientSet, err := kubernetes.NewForConfig(config)
+		if err != nil {
+			return fmt.Errorf("error initializing kubernetes client: %w", err)
+		}
+
+		kubeClient, _, err := kube.NewKubeHTTPClient()
+		if err != nil {
+			return fmt.Errorf("failed to create kube client: %w", err)
+		}
+
+		ctx := context.Background()
+		_, gitProvider, err := factory.GetGitClients(ctx, kubeClient, providerClient, services.GitConfigParams{
+			ConfigRepo:       opts.ConfigRepo,
+			Namespace:        opts.Namespace,
+			IsHelmRepository: true,
+			DryRun:           false,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to get git clients: %w", err)
+		}
+
+		profilesService := profiles.NewService(clientSet, log)
+		return profilesService.Add(ctx, gitProvider, opts)
+	}
+}
+
+func validateAddOptions(opts profiles.AddOptions) error {
```

Comment:
This sets up an unfortunate error pattern, where you run it and it tells you something is invalid, so you fix it, and then you run it again, and it tells you something _else_ is invalid.

Maybe a [`multierror`](https://github.com/hashicorp/go-multierror) would be better?

At least, return a slice of strings with all the errors.

Reply:
Isn't the problem that these depend on each other? If `opts.Name` is not defined, you can't call `ApplicationNameTooLong`. And then you can't call HasPrefix either. Though I agree that ConfigRepo and Cluster could be checked separately.

Reply:
Changed this in the commit that addresses review feedback -  the `validateAddOptions` func only has 2 checks in it now since I moved the required flags to Cobra's feature for checking for required flags.